tests/testthat/test_gapfilling.R

```diff
@@ -0,0 +1,395 @@
+    pointblank::expect_col_vals_in_set(
+      source_value,
+      c(
+        "First value carried backwards",
+        "Original",
+        "Last value carried forward"
+      )
+    )
+})
+
+# proxy_fill --------------------------------------------------------------------
+
+testthat::test_that("proxy_fill scales gaps from proxy ratios", {
+  result <- proxy_fill(
+    proxy_fill_fixture(),
+    value,
+    proxy_variable,
+    year,
+    .by = "category"
+  )
+
+  result |>
+    pointblank::expect_col_exists("proxy_ratio") |>
+    pointblank::expect_col_exists("source_value") |>
+    pointblank::expect_col_vals_in_set(
+      source_value,
+      c(
+        "Original",
+        "Proxy interpolated",
+        "Proxy carried forward",
+        "Proxy carried backwards"
+      )
+    )
+
+  result |>
+    dplyr::filter(source_value == "Original") |>
+    dplyr::pull(value) |>
+    testthat::expect_equal(c(3, 0, 1, 5))
+
+  ratio_diff <- result |>
+    dplyr::filter(!is.na(value)) |>
+    dplyr::mutate(calculated_ratio = value / proxy_variable) |>
+    dplyr::summarise(
+      max_diff = max(abs(calculated_ratio - proxy_ratio), na.rm = TRUE)
+    ) |>
+    dplyr::pull(max_diff)
+
+  testthat::expect_lt(ratio_diff, 1e-6)
+
+  dplyr::is_grouped_df(result) |>
+    testthat::expect_false()
+})
+
+testthat::test_that("proxy_fill works without grouping variables", {
+  simple_proxy <- tibble::tribble(
+    ~year, ~value, ~proxy_variable,
+    2015, 10, 5,
+    2016, NA, 10,
+    2017, 30, 15
+  )
+
+  proxy_fill(simple_proxy, value, proxy_variable, year) |>
+    pointblank::expect_col_exists("proxy_ratio") |>
+    pointblank::expect_col_vals_not_null(proxy_ratio)
+})
+
+# sum_fill ----------------------------------------------------------------------
+
+testthat::test_that("sum_fill accumulates changes while keeping originals", {
+  result <- sum_fill(
+    sum_fill_fixture(),
+    value,
+    change_variable,
+    start_with_zero = TRUE,
+    .by = "category"
+  )
+
+  result |>
+    pointblank::expect_col_exists("source_value") |>
+    pointblank::expect_col_vals_in_set(
+      source_value,
+      c("Original", "Filled with sum")
+    ) |>
+    pointblank::expect_col_vals_not_null(value)
+
+  result |>
+    dplyr::filter(category == "a") |>
+    dplyr::pull(value) |>
+    testthat::expect_equal(c(2, 5, 3, 6, 10, 0, 1))
+
+  result |>
+    dplyr::filter(category == "b") |>
+    dplyr::pull(value) |>
+    testthat::expect_equal(c(1, 1, 1, 1, 5, 6))
+})
+
+testthat::test_that("sum_fill handles accumulation without explicit groups", {
+  simple_series <- tibble::tribble(
+    ~year, ~value, ~change_variable,
+    2015, 10, 0,
+    2016, NA, 2,
+    2017, NA, 3,
+    2018, NA, 1
+  )
+
+  result <- sum_fill(simple_series, value, change_variable)
+
+  result |>
+    pointblank::expect_col_vals_equal(value, c(10, 12, 15, 16)) |>
+    pointblank::expect_col_vals_in_set(
+      source_value,
+      c("Original", "Filled with sum")
+    )
```

Comment:
Again probably just a matter of style, but I like writing the pipeline directly:
```suggestion
  simple_series |>
    sum_fill(value, change_variable) |>
    pointblank::expect_col_vals_equal(value, c(10, 12, 15, 16)) |>
    pointblank::expect_col_vals_in_set(
      source_value,
      c("Original", "Filled with sum")
    )
```